### PR TITLE
Feat/e2e

### PR DESCRIPTION
**Pull request by yaniszuck**
Ci for Kubi feat/e2e

### CODE REVIEW

File: .github/Csr.sh [hunk 1]
```diff
@@ -0,0 +1,168 @@
+#!/bin/bash
+
+
+# kubi private key 
+openssl ecparam -genkey -name secp521r1 -noout -out /tmp/ecdsa-key.pem
```
Comment (evrardjp) [inline .github/Csr.sh:5]: it would probably be better to build a folder from mktemp

File: .github/Csr.sh [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+# kubi public key
+openssl ec -in /tmp/ecdsa-key.pem -pubout -out /tmp/ecdsa-public.pem
+
+
+mkdir dev
+cd dev
+
+# recuperer les adresses ip des nodes
```
Comment (evrardjp) [inline .github/Csr.sh:14]: english?

File: .github/Csr.sh [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+#creation du certificat pour  faire la demande de signature par kube CA
+
+# install CFSSL tools
+
+VERSION=$(curl --silent "https://api.github.com/repos/cloudflare/cfssl/releases/latest" | grep '"tag_name"' | sed -E 's/.*"([^"]+)".*/\1/')
+VNUMBER=${VERSION#"v"}
+wget https://github.com/cloudflare/cfssl/releases/download/${VERSION}/cfssljson_${VNUMBER}_linux_amd64 -O cfssljson
+chmod +x cfssljson
+sudo mv cfssljson /usr/local/bin
```
Comment (evrardjp) [inline .github/Csr.sh:30]: you can replace both the chmod  +x and sudo mv with
`sudo install `

File: .github/Csr.sh [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+cfssljson -version
+VERSION=$(curl --silent "https://api.github.com/repos/cloudflare/cfssl/releases/latest" | grep '"tag_name"' | sed -E 's/.*"([^"]+)".*/\1/')
```
Comment (evrardjp) [inline .github/Csr.sh:33]: Isn't this the same as VERSION above? Why doing it twice?

File: .github/Csr.sh [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+VNUMBER=${VERSION#"v"}
+wget https://github.com/cloudflare/cfssl/releases/download/${VERSION}/cfssl_${VNUMBER}_linux_amd64 -O cfssl
+chmod +x cfssl
+sudo mv cfssl /usr/local/bin
```
Comment (evrardjp) [inline .github/Csr.sh:37]: ditto

File: .github/Csr.sh [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+  "key": {
+    "algo": "rsa",
+    "size": 2048
+  }
+}
+EOF
+
+
+
+echo '{
```
Comment (evrardjp) [inline .github/Csr.sh:92]: maybe this can be a static file in the tests/ folder ?

File: .github/Csr.sh [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+
+cat <<EOF | cfssl gencert -initca - | cfssljson -bare ca
+{
+  "CN": "kube-chaos",
+  "key": {
+    "algo": "rsa",
+    "size": 2048
+  }
+}
```
Comment (evrardjp) [inline .github/Csr.sh:87]: maybe this can be a fixture/static file in the tests/ folder?

File: .github/Csr.sh [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+VNUMBER=${VERSION#"v"}
+wget https://github.com/cloudflare/cfssl/releases/download/${VERSION}/cfssl_${VNUMBER}_linux_amd64 -O cfssl
+chmod +x cfssl
+sudo mv cfssl /usr/local/bin
+
+# create certificat 
+cat <<EOF | cfssl genkey - | cfssljson -bare server
+{
+  "hosts": [
+    "$PUBLIC_APISERVER_URL",
```
Comment (evrardjp) [inline .github/Csr.sh:43]: Wouldn't the PUBLIC_APISERVER_URL be always the same with kind?
Would it make sense to have this as static file under tests/ ?

File: .github/Csr.sh [hunk 9]
```diff
@@ -0,0 +1,168 @@
+#!/bin/bash
+
+
+# kubi private key 
+openssl ecparam -genkey -name secp521r1 -noout -out /tmp/ecdsa-key.pem
+
+# kubi public key
```
Comment (evrardjp) [inline .github/Csr.sh:7]: I probably would generate these keys staticly for CI, and use always the same cert, to minimize surprises...

File: .github/end_to_end/y_test.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
+package main
```
Comment (evrardjp) [inline .github/end_to_end/y_test.go:1]: why is it named y_test? 

why didn't you include this in tests/ directly ?

File: .github/end_to_end/y_test.go [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+		return nil, err
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return clientset, nil
+}
+
+// function to check existing namespace
```
Comment (evrardjp) [inline .github/end_to_end/y_test.go:32]: isn't that comment a bit obvious? ;)

File: .github/workflows/chaos-experiment.yaml [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+               KUBERNETES_SERVICE_PORT="API_PORT" \
+               LDAP_PASSWD="password" \
+               go run ./cmd/api//main.go &
+     
+              
+      - name: existing  namespaces test 
+        run:  | 
+               go get k8s.io/client-go/tools/clientcmd@v0.20.4
+               sleep 10m
+               go test -v  ./.github/end_to_end/y_test.go
```
Comment (evrardjp) [inline .github/workflows/chaos-experiment.yaml:78]: I don't see a big no to do it this way but I can't shake the feeling it doesn't seem right.
I am worried that we might have tests later, and `go test` done recursively will start to break because the go test isn't implemented with a go binary that loads the fixture (fixture is loaded earlier). 
Because function and test is not shipped together, I would rather have a simple shell script that does the equivalent. What do you think?

File: cmd/operator/main.go [hunk 13]
```diff
@@ -42,8 +42,11 @@ func main() {
 		utils.Log.Info().Msg("NetworkPolicies generation is disabled.")
 	}
 	services.WatchProjects()
-
-	timerKubiRefresh := time.NewTicker(10 * time.Minute)
+        
+	//os.Setenv("timer_refresh" , "10")
+	timerKubiRefresh := time.NewTicker(5 * time.Minute)
```
Comment (evrardjp) [inline cmd/operator/main.go:47]: I don't think it's the right idea to change the default value in the implementation of an e2e test.

It makes more sense to me to add a variable, which defaults to current value of 10 minutes, and override it for the test.

File: .github/workflows/chaos-experiment.yaml [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+      - name: Set up Go
+        uses: actions/setup-go@v2
+        with:
+          go-version: 1.17
+          go-path: ${{ github.workspace }}/go
+          
+      - name: Set GOROOT
+        run: echo "export GOROOT=/opt/hostedtoolcache/go/1.17.13/x64" >> $GITHUB_ENV 
+          
+      - name: Create certificate for kubi
```
Comment (evrardjp) [inline .github/workflows/chaos-experiment.yaml:41]: For a test, maybe you can just ship a default set of certificates and values, so that the test is relatively cheaper, faster to load, and won't need to fetch stuff from the external world? This would make the life simpler.

File: .github/workflows/chaos-experiment.yaml [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+      - name: Create cluster KinD
+        uses: helm/kind-action@v1.4.0
+        with:
+          config: .github/kind-cluster-${{ matrix.kubernetes }}.yaml
+      - name: testing cluster kinD
+        run: kubectl cluster-info --context kind-chart-testing echo " current-context:"
+          $(kubectl config current-context)
+                 
+      - name: Setting up kubeconfig ENV for Github Chaos Action
+        run: echo ::set-env name=KUBE_CONFIG_DATA::$(base64 -w 0 ~/.kube/config)
```
Comment (evrardjp) [inline .github/workflows/chaos-experiment.yaml:26]: isn't ::set-env deprecated?

File: .github/kind-cluster-1.23.yaml [hunk 16]
```diff
@@ -0,0 +1,12 @@
+kind: Cluster
```
Comment (evrardjp) [inline .github/kind-cluster-1.23.yaml:1]: I don't think you need to include 1.23 yet, as this commit's purpose is only to test e2e. Hence, it doesn't add support of a new version. 

If you want to try supporting 1.23, I suggest you to split that into its own commit.

File: .github/Csr.sh (hunk 1 above)
Comment (yaniszuck) [reply .github/Csr.sh:5]: done !

File: .github/Csr.sh (hunk 2 above)
Comment (yaniszuck) [reply .github/Csr.sh]: done !

File: .github/Csr.sh (hunk 3 above)
Comment (yaniszuck) [reply .github/Csr.sh:24]: sudo Install is not working !! ( packages not found )

File: cmd/operator/main.go (hunk 13 above)
Comment (yaniszuck) [reply cmd/operator/main.go:47]: done !

File: .github/workflows/chaos-experiment.yaml [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+      - name: Checkout code
+        uses: actions/checkout@v2
+      - name: Set up Go
+        uses: actions/setup-go@v2
+        with:
+          go-version: 1.17
+          go-path: ${{ github.workspace }}/go
+          
+      - name: Set GOROOT
+        run: echo "export GOROOT=/opt/hostedtoolcache/go/1.17.13/x64" >> $GITHUB_ENV
```
[the remark this answers is not in the document]
Comment (yaniszuck) [reply .github/workflows/chaos-experiment.yaml]: done !

File: .github/kind-cluster-1.23.yaml (hunk 16 above)
Comment (yaniszuck) [reply .github/kind-cluster-1.23.yaml:1]: done !
